README.md
**THIS PACKAGE IS NOT MAINTAINED ANYMORE**

# Check all links in a Laravel app

[![Latest Version on Packagist](https://img.shields.io/packagist/v/spatie/laravel-link-checker.svg?style=flat-square)](https://packagist.org/packages/spatie/laravel-link-checker)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)
[![Build Status](https://img.shields.io/travis/spatie/laravel-link-checker/master.svg?style=flat-square)](https://travis-ci.org/spatie/laravel-link-checker)
[![StyleCI](https://github.styleci.io/repos/45639655/shield?branch=master)](https://github.styleci.io/repos/45639655)
[![Quality Score](https://img.shields.io/scrutinizer/g/spatie/laravel-link-checker.svg?style=flat-square)](https://scrutinizer-ci.com/g/spatie/laravel-link-checker)
[![Total Downloads](https://img.shields.io/packagist/dt/spatie/laravel-link-checker.svg?style=flat-square)](https://packagist.org/packages/spatie/laravel-link-checker)

This package provides a command that can check all links on your laravel app. By default, it will log all
links that do not return a status code in the 200- or 300-range. There's also an option to mail broken links.

If you like this package, take a look at [the other ones we have made](https://spatie.be/opensource/laravel).

## Install

You can install the package via composer:
``` bash
composer require spatie/laravel-link-checker
```

Next, you must install the service provider:

```php
// config/app.php
'providers' => [
    ...
    Spatie\LinkChecker\LinkCheckerServiceProvider::class,
];
```

The service provider will automatically be registered.

You can optionally publish the config-file with:
```bash
php artisan vendor:publish --provider="Spatie\LinkChecker\LinkCheckerServiceProvider" --tag="config"
```

This is the contents of the published config file:

```php
return [

    /*
     * The base url of your app. Leave this empty to use
     * the url configured in config/app.php
     */
    'url' => '',

    /*
     * The profile determines which links need to be checked.
     */
    'default_profile' => Spatie\LinkChecker\CheckAllLinks::class,

    /*
     * The reporter determines what needs to be done when the
     * the crawler has visited a link.
     */
    'default_reporter' => Spatie\LinkChecker\Reporters\LogBrokenLinks::class,
    
    /*
     * To speed up the checking process we'll fire off requests concurrently.
     * Here you can change the amount of concurrent requests.
     */
    'concurrency' => 10

    /*
     *  Here you can specify configuration regarding the used reporters
     */
    'reporters' => [

        'mail' => [

            /*
             * The `from` address to be used by the mail reporter.
             */
            'from_address' => '',
            
            /*
             * The `to` address to be used by the mail reporter.
             */
            'to_address' => '',

            /*
             * The subject line to be used by the mail reporter.
             */
            'subject' => '',
        ],

        /*
         * If you wish to exclude status codes from the reporters,
         * you can select the status codes that you wish to
         * exclude in the array below like: [200, 302]
         */
        'exclude_status_codes' => [],
    ],
];
```


## Usage

You can start checking all links by issuing this command:

```bash
php artisan link-checker:run
```

Want to run the crawler on a different url? No problem!

```bash
php artisan link-checker:run --url=https://laravel.com
```


### Schedule the command 
To frequently check all links you can schedule the command:

```php
// app/console/Kernel.php

protected function schedule(Schedule $schedule)
{
    ...
    $schedule->command('link-checker:run')->sundays()->daily();
}
``` 

### Mail broken links
By default the package will log all broken links. If you want to have them mailed instead, just specify
`Spatie\LinkChecker\Reporters\MailBrokenLinks` in the `default_reporter` option in the config file.

## Creating your own crawl profile
A crawlprofile determines which links need to be crawled. By default `Spatie\LinkChecker\CheckAllLinks` is used,
which will check all links it finds. This behaviour can be customized by specifying a class in the `default_profile`-option in the config file.
The class must extend the abstract class `Spatie\Crawler\CrawlProfile`:

```php
abstract class CrawlProfile
{
    /**
     * Determine if the given url should be crawled.
     *
     * @param \Psr\Http\Message\UriInterface $url
     *
     * @return bool
     */
    abstract public function shouldCrawl(UriInterface $url): bool;
}
```

## Creating your own reporter
A reporter determines what should be done when a link is crawled and when the crawling process is finished.
This package provides two reporters: `Spatie\LinkChecker\Reporters\LogBrokenLinks` and `Spatie\LinkChecker\Reporters\MailBrokenLinks`.
You can create your own behaviour by making a class extend the abstract class `Spatie\Crawler\CrawlObserver`:

```php
abstract class CrawlObserver
{
    /**
     * Called when the crawler will crawl the url.
     *
     * @param \Psr\Http\Message\UriInterface $url
     */
    public function willCrawl(UriInterface $url)
    {
    }

    /**
     * Called when the crawler has crawled the given url successfully.
     *
     * @param \Psr\Http\Message\UriInterface $url
     * @param \Psr\Http\Message\ResponseInterface $response
     * @param \Psr\Http\Message\UriInterface|null $foundOnUrl
     */
    abstract public function crawled(
        UriInterface $url,
        ResponseInterface $response,
        ?UriInterface $foundOnUrl = null
    );

    /**
     * Called when the crawler had a problem crawling the given url.
     *
     * @param \Psr\Http\Message\UriInterface $url
     * @param \GuzzleHttp\Exception\RequestException $requestException
     * @param \Psr\Http\Message\UriInterface|null $foundOnUrl
     */
    abstract public function crawlFailed(
        UriInterface $url,
        RequestException $requestException,
        ?UriInterface $foundOnUrl = null
    );

    /**
     * Called when the crawl has ended.
     */
    public function finishedCrawling()
    {
    }
}
``` 
  
To make it easier to create a reporter, you can extend `Spatie\LinkChecker\Reporters\BaseReporter` which
provides many useful methods.

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Testing

First, start the test server in a separate terminal session:
``` bash
cd tests/server
./start_server.sh
``` 

With the server running you can execute the tests
``` bash
composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please email freek@spatie.be instead of using the issue tracker.

## Postcardware

You're free to use this package (it's [MIT-licensed](LICENSE.md)), but if it makes it to your production environment we highly appreciate you sending us a postcard from your hometown, mentioning which of our package(s) you are using.

Our address is: Spatie, Samberstraat 69D, 2060 Antwerp, Belgium.

All postcards are published [on our website](https://spatie.be/en/opensource/postcards).

## Credits

- [Freek Van der Herten](https://github.com/freekmurze)
- [All Contributors](../../contributors)

## Support us

Spatie is a webdesign agency based in Antwerp, Belgium. You'll find an overview of all our open source projects [on our website](https://spatie.be/opensource).

Does your business depend on our contributions? Reach out and support us on [Patreon](https://www.patreon.com/spatie). 
All pledges will be dedicated to allocating workforce on maintenance and new awesome stuff.

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `laravel-link-checker` will be documented in this file

## 4.2.0 2009-11-26

- remove support for older php versions
- remove support for older laravel versions
- support laravel 5.8 and 6

## 4.1.1 - 2018-08-27

- add support for Laravel 5.7
- fix mail when broken urls are found

## 4.1.0 - 2018-04-26

- add client options to config file

## 4.0.0 - 2018-03-20

- Update to `spatie/crawler ^4.0`
- Drop `illuminate/support ~5.5.0` support

## 3.0.0 - 2018-02-12

- Update to `spatie/crawler ^3.0`
- Support Laravel 5.5 and 5.6

## 2.4.0 - 2018-01-17
- add `exclude_status_codes`

## 2.3.0 - 2018-01-17
- add "found on" in the `MailBrokenLinks` reporter

## 2.2.2 - 2017-10-24
- fix mail

## 2.2.1 - 2017-10-14
- permissions fix

## 2.2.0 - 2017-09-29
- add subject line to email

## 2.1.0 - 2017-09-29
- fix incorrect tagged versions
- improve speed by crawling the app with multiple concurrent connections
- add autodiscovery for Laravel 5.5

## 1.1.0 - 2017-06-29
**THIS VERSION WAS TAGGED INCORRECTLY, DO NOT USE**

- add autodiscovery for Laravel 5.5

## 1.0.5 - 2017-06-29
**THIS VERSION WAS TAGGED INCORRECTLY, DO NOT USE**
- remove hidden `laravel-collective/html` dependency 

## 1.0.4 - 2017-01-19
**THIS VERSION WAS TAGGED INCORRECTLY, DO NOT USE**

- remove another typehint on a protected method

## 1.0.3 - 2017-01-19
**THIS VERSION WAS TAGGED INCORRECTLY, DO NOT USE**

- remove typehint from `isSuccessOrRedirect`

## 1.0.2 - 2016-12-03
**THIS VERSION WAS TAGGED INCORRECTLY, DO NOT USE**

- improve speed by crawling the app with multiple concurrent connections

## 1.0.1 - 2015-12-03

- The mail reporter now only reports broken urls

## 1.0.0 - 2015-11-25

- Initial release


resources/config/laravel-link-checker.php
<?php

return [

    /*
     * The base url of your app. Leave this empty to use
     * the url configured in config/app.php.
     */
    'url' => '',

    /*
     * The profile determining which links need to be checked.
     */
    'default_profile' => Spatie\LinkChecker\CheckAllLinks::class,

    /*
     * The reporter determining what needs to be done when the
     * the crawler has visited a link.
     */
    'default_reporter' => Spatie\LinkChecker\Reporters\LogBrokenLinks::class,

    /*
     * To speed up the checking process we'll fire off requests concurrently.
     * Here you can change the amount of concurrent requests.
     */
    'concurrency' => 10,

    /*
     * List of client options to use in Guzzle HTTP.
     */
    'client_options' => [

        \GuzzleHttp\RequestOptions::COOKIES => true,
        \GuzzleHttp\RequestOptions::CONNECT_TIMEOUT => 10,
        \GuzzleHttp\RequestOptions::TIMEOUT => 10,
        \GuzzleHttp\RequestOptions::ALLOW_REDIRECTS => false,

    ],

    /*
     * Configuration regarding the used reporters.
     */
    'reporters' => [

        'mail' => [

            /*
             * The from address to be used by the mail reporter.
             */
            'from_address' => '',

            /*
             * The to address to be used by the mail reporter.
             */
            'to_address' => '',

            /*
             * The subject line to be used by the mail reporter.
             */
            'subject' => '',

        ],

        /*
         * If you wish to exclude certain status codes from the reporters,
         * specify them here.
         * Here's an example: [200, 302]
         */
        'exclude_status_codes' => [],
    ],
];
